packages/planner_node/src/astar.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include "model/params.h"
+#include "collision_checker/collision_checker.h"
+
+#include <tf2/impl/utils.h>
+
+#include <cmath>
+#include <vector>
+#include <string>
+#include <fstream>
+#include <sstream>
+
+namespace truck::astar {
+
+/**
+ * Defining path elementary parts called primitives
+ * @details each primitive is described by a set of subpoints (x,y) and directions in them (yaw)  
+ */
+struct Primitive {
+    std::vector<geom::Vec2> sub_points;
+    std::vector<double> yaws;
+};
+
+/**
+ * Defining 2D indices of graph's vertices
+ * @details each vertex has 2 indices (i, j) which are limited
+ * @details by the physical dimensions of the graph and its resolution 
+ */
+struct Index2D {
+    int i, j;
+};
+
+/**
+ * Defining graph vertices
+ */
+struct Vertex {
+    // position info
+    Index2D index_;
+    geom::Vec2 coord_;
+    double incoming_yaw_;
+
+    // scores info
+    double g_, h_, f_;
+
+    // neighbors info
+    std::vector<Vertex*> neighbors_;
+    std::vector<int> neighbors_primitives_indices_;
+
+    // previous vertex pointer
+    Vertex* prev_vertex_{nullptr};
+
+    // index of primitive connecting previous and current vertices
+    int prev_vertex_primitive_index_{NULL};
```

Comment:
Вот это на самом деле не вершина, а состояние поиска, это нужно отделить.

Reply:
Короче, я не выдержал и набросал архитектурку...

```
namespace truck::planner::search {

// So we introduce a abstraction of "grid".
// It is a collection of nodes (point and id).
// Any edge of graph starts from one node and ends in another (but yaw may be different).
using NodeId = size_t;
using NodeIds = std::vector<NodeId>;

struct Node {
  NodeId id;
  geom::Point point;
  bool is_finish; // some nodes are finish (at first step let's ignore direction)
};



// Than we split building process and usage of grid.
struct GridParams { /* you params here */ };

class GridBuilder {
public:
    GridBuilder(const GridParams& params);

    // It's a common hack, that then allows you to write:
    // auto builder = GridBuilder(params)
    //    .SetEgoPose(ego_pose)
    //    .SetStaticCollisionChecker(static_collision_checker);
    GridBuilder& SetEgoPose(const geom::Point ego_pose);
    GridBuilder& SetFinisArea(const geom::Circle area);
    GridBuilder& SetStaticCollisionChecker(std::shared_ptr<StaticCollisionChecker> static_collision_checker);

    Grid Build() const {
        // Some fields mey be optional. e.g. static_collision_checker_.
        // But it's better to drop  nodes in collision, using static_collision_checker_.
    }

private:
    std::optional<geom::Point> ego_pose_;
    std::optional<geom::Circle> finish_area_;
    std::shared_ptr<StaticCollisionChecker> static_collision_checker_;
};


class Grid {

public:
    // Inside can be written a lot of auxiliary code...
    const Node& GetNodeById(NodeId node_id) const;
    NodeIds GetNeighboors(NodeId node_id) const;

    NodeId GetEgoNode() const;

private:
  // It's often convenient to use index in array as id :)
  std::vector<Node> nodes_;
};

// So, you may write some code for yaw binning.
template<size_t yaws_count>
class YawBins {

public:
    // use snap here
    size_t GetIndex(geom::Angle yaw) const;
    size_t GetIndex(geom::Vector direction) const;

    geom::Angle GetAngle(size_t index) const;
    geom::Vector GetDirection(size_t index) const;

public:
    std::array<geom::Vector, yaws_count> directions_;
    std::array<geom::Angle, yaws_count> angles_;
};

private:
  static std::array<geom::Vector, yaws_count> yaws_;
};


// It's good idea for current regular grid (may changed in future).
struct EdgeGeometryId {
    size_t offset_x;
    size_t offset_y;
    size_t from_yaw;
    size_t to_yaw;
};

class EdgeGeometryCache {
public:
    // It's not a const method, because it may change cache.
    const Poses &GetEdgeGeometry(const EdgeGeometryId& index);

private:
    std::map<EdgeGeometryId, Poses> cache_;
};

struct Vertex;

struct Edge {
    const VertexId from;
    const VertexId to;
    EdgeGeometryId primitive_id;
};

// Next we introduce a vertex abstraction, it's a node with yaw index (fix a truck position).
using VertexId = size_t;

struct Vertex {
    VertexId id;
    NodeId node_id;
    size_t yaw_index;
    VecItemPtr<Edge> edges; // edges without collisions
};

template<T>
class VecItemPtr {

public:
    T& operator*() { return (*item)[index_]; }
    T* operator->() { return &(*item)[index_]; }

private:
    size_t index_;
    std::vector<T> *items;
}


// Graph is based on grid and
class DynamicGraph {

public:
    DynamicGraph(const GraphParams& params);
    const Vertex& GetVertex(VertexId vertex_id) const {
        // check that edges are initialized
    }

    const Vertex& GetStartVertex() const;

private:
    std::shared_ptr<Grid> grid_;
    // to check edge collision
    std::shared_ptr<DynamicCollisionChecker> dynamic_collision_checker_;
    std::shared_ptr<EdgeGeometryCache> edge_geometry_cache_;
    YawBins yaw_bins_;

    struct Cache {
        std::vector<Vertex> vertices;
        std::vector<bool> initialized;
        std::vector<Edge> edges;
    } cache_;
};

// for each vertex we know
struct VertexSearhState {
    Vertex* prev_vertex;
    EdgePtr edge_ptr;
};

class Result {
    std::vector<VertexId> vertex_ids;
    std::vector<EdgePtr> edge_ptrs;

    std::vector<Pose> GetTrace() const;
};

class Searcher {

public:

    Searcher(...);

    Result FindPath() const;

    // setters here

private:
    std::shared_ptr<DynamicGraph> graph_;
};
```